### ISSUE TITLE
AJAX problems in latest 3.1.0 using TokenPerPage

### ISSUE THREAD

**Issue by cxzzero** [aramrami/OWASP-CSRFGuard#57]
Thanks in advance for view this post.

I am facing problems like what is being posted in [this link in StackOverflow](http://stackoverflow.com/questions/31390625/csrfguard-how-to-inject-csrf-token-into-url-returned-by-ajax-call):

I see the last commit on project was done 17 Dec 2015 and that post is earlier, but as I see error remains and I can not make it happen.

I will try to explain my best:

I am adding CSRF defenses to my site: 
- added both techniques Javascript DOM manipulation and JSP Tag configuration as explained in [OWASP CSRF Guard manual](https://www.owasp.org/index.php/CSRFGuard_3_User_Manual)
- changed csrfguard property to only protect selected pages (I decided that for testing purposes to reduce to minimum issues).
- The sample below focuses on JSP Tag but I also tested how it behaves using Javascript DOM manipulation mode and reached the same results.
- unique TokenPerPage enabled

Say I java a jsp page like this:

```
[...]
<%@ taglib uri="/WEB-INF/csrfguard.tld" prefix="csrf" %>
[...]
<form action="a1Action.jsp" method="post">
      [...]
      <input type="hidden" name="<csrf:tokenname/>" value="<csrf:tokenvalue uri="a1Action.jsp"/>" />
</form>
[...]
```

On Ajax Request a backend function is called in _a1Action.jsp_ and then a dialog appears when ok or error, but page a1.jsp remains in view, so it is possible to click submitt button again.

**The behaviour I saw in code is**:
- The first time I click submit button, behaviour is OK (works like a charm!): the unique token calculated from server side and from request are the same, so protected resource is received and backend function called. Great.
- The thing is that after submit button is clicked again, protected resource is never called, because token per page sent in request does not get refreshed, but server side token per page was changed, so it is detected like a CSRF attack. I see these things:
- csrf token remains unchanged in input hidden after AJAX call
- csrf token per page server side is calculated again and because of that the line **!tokenFromPages.equals(tokenFromRequest)** in code that follows is happening (it is weird because it shoud not come to that place, it should go to **verifyAjaxToken()** method, but as I see **X-Requested-With** is **not added to HTTP request** (tested with Fiddler and debbuging code):

```
private void verifyPageToken(HttpServletRequest request)
       [...]
           } else if (tokenFromPages != null) {
            if (!tokenFromPages.equals(tokenFromRequest)) {
                /** FAIL: request does not match page token **/
                throw new CsrfGuardException("request token does not match page token");
            }
       [...]
}
```

To summarize, because of that subsequent calls to submit button shows like a potential cross-site request forgery (CSRF) attack. 

Could you help me in an acceptable solution I can do to solve the problem? Is there any configuration missed from me in jsp? How about the solution brought from post in StackOverflow?

Regards and thanks,
Juan Pablo

**Comment by cxzzero** [comment]
I made a lot more of research about the issue and found out:

- an old post from Eric Sheridan (dated on [December 2010](http://ericsheridan.blogspot.com.br/2010/12/how-csrfguard-protects-ajax.html)) saying:
_"I have not yet figured out a clean solution to support per-page tokens within Ajax requests which would not either break application functionality or increase network traffic two-fold."_
**This post is earlier than the release of version 3, but could be valid anyway.**

- Default **csrfguard.properties** still say that _"Use of the unique token per page property is currently experimental"_

Just to update the issue and keep track of it.
